pkg/cassandra/auth.go

```diff
@@ -0,0 +1,113 @@
+package cassandra
+
+import (
+	"fmt"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/images"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+)
+
+const JmxInitContainer = "jmx-credentials"
+
+// ApplyAuth modifies the dc config depending on whether auth is enabled in the cluster or not.
+func ApplyAuth(dcConfig *DatacenterConfig, authEnabled bool) {
+
+	dcConfig.CassandraConfig = ApplyAuthSettings(dcConfig.CassandraConfig, authEnabled)
+
+	// By default, the Cassandra process will be started with LOCAL_JMX=yes, see cassandra-env.sh. This means that the
+	// Cassandra process will only be accessible with JMX from localhost. This is the safest and preferred setup: you
+	// still can use JMX by SSH'ing into the Cassandra pod, for example to run nodetool. But some components need remote
+	// JMX access (Reaper, metrics, etc.). Such components and their controllers are responsible for setting
+	// LOCAL_JMX=no whenever appropriate, to enable remote JMX access.
+	// However, authentication will get in the way, even if it's an orthogonal concern. Indeed, with LOCAL_JMX=yes
+	// cassandra-env.sh will infer that no JMX authentication should be used
+	// (com.sun.management.jmxremote.authenticate=false), whereas with LOCAL_JMX=no it will infer that authentication is
+	// required (com.sun.management.jmxremote.authenticate=true). We need to change that here and enable/disable
+	// authentication based on what the user specified, not what the script infers. The only way to achieve this is by
+	// setting the JVM_EXTRA_OPTS env var here to contain the JVM option -Dom.sun.management.jmxremote.authenticate.
+	// Indeed, the JVM_EXTRA_OPTS env var is appended to the JVM_OPTS var at the very end of the cassandra-env.sh
+	// script, which means that any value set there will override what the script itself had previously computed.
+	// FIXME this is a bit hacky and might interfere with the way cass-operator enables DSE workloads
+	if dcConfig.PodTemplateSpec == nil {
+		dcConfig.PodTemplateSpec = &corev1.PodTemplateSpec{}
+	}
+	UpdateCassandraContainer(dcConfig.PodTemplateSpec, func(c *corev1.Container) {
+		jmxAuthenticateOpt := fmt.Sprintf("-Dcom.sun.management.jmxremote.authenticate=%v", authEnabled)
+		c.Env = append(c.Env,
+			corev1.EnvVar{
+				Name:  "JVM_EXTRA_OPTS",
```

Comment:
Can you write up a ticket for making `JVM_EXTRA_OPTS` configurable. I don't think this is something we will get done for 1.0.0, but it would be a nice enhancement. There is a lot in flight right now, and I don't want it to fall through the cracks.

Reply:
In fact there is a way to achieve what we want without resorting to `JVM_EXTRA_OPTS`: in config-builder, the cassandra-env.sh template also accepts an `additional-jvm-opts` parameter. I suggest that we use it instead, I will make the necessary changes.

Reply:
Done here: https://github.com/k8ssandra/k8ssandra-operator/blob/0d829b9246e62870d9345ada8660153a049bb6f8/pkg/cassandra/config.go#L71-L78.